## Import GraphLab & Data

In [1]:
import graphlab

In [2]:
sales = graphlab.SFrame('kc_house_data.gl/')

[INFO] 1449409502 : INFO:     (initialize_globals_from_environment:282): Setting configuration variable GRAPHLAB_FILEIO_ALTERNATIVE_SSL_CERT_FILE to /Users/rogeramitchell/.graphlab/anaconda/lib/python2.7/site-packages/certifi/cacert.pem
1449409502 : INFO:     (initialize_globals_from_environment:282): Setting configuration variable GRAPHLAB_FILEIO_ALTERNATIVE_SSL_CERT_DIR to 
This non-commercial license of GraphLab Create is assigned to rogeramitchell@gmail.com and will expire on October 21, 2016. For commercial licensing options, visit https://dato.com/buy/.

[INFO] Start server at: ipc:///tmp/graphlab_server-22048 - Server binary: /Users/rogeramitchell/.graphlab/anaconda/lib/python2.7/site-packages/graphlab/unity_server - Server log: /tmp/graphlab_server_1449409502.log
[INFO] GraphLab Server Version: 1.7.1


## Split data for training & testing

In [3]:
train_data, test_data = sales.random_split(.8, seed = 0)

In [4]:
train_data

Columns:
	id	str
	date	datetime
	price	float
	bedrooms	float
	bathrooms	float
	sqft_living	float
	sqft_lot	int
	floors	str
	waterfront	int
	view	int
	condition	int
	grade	int
	sqft_above	int
	sqft_basement	int
	yr_built	int
	yr_renovated	int
	zipcode	str
	lat	float
	long	float
	sqft_living15	float
	sqft_lot15	float

Rows: Unknown

Data:
+------------+---------------------------+-----------+----------+-----------+
|     id     |            date           |   price   | bedrooms | bathrooms |
+------------+---------------------------+-----------+----------+-----------+
| 7129300520 | 2014-10-13 00:00:00+00:00 |  221900.0 |   3.0    |    1.0    |
| 6414100192 | 2014-12-09 00:00:00+00:00 |  538000.0 |   3.0    |    2.25   |
| 5631500400 | 2015-02-25 00:00:00+00:00 |  180000.0 |   2.0    |    1.0    |
| 2487200875 | 2014-12-09 00:00:00+00:00 |  604000.0 |   4.0    |    3.0    |
| 1954400510 | 2015-02-18 00:00:00+00:00 |  510000.0 |   3.0    |    2.0    |
| 7237550310 | 2014-05-12 00:00:00+00:00 | 1225000.0 |   4.0    |    4.5    |
| 1321400060 | 2014-06-27 00:00:00+00:00 |  257500.0 |   3.0    |    2.25   |
| 2008000270 | 2015-01-15 00:00:00+00:00 |  291850.0 |   3.0    |    1.5    |
| 2414600126 | 2015-04-15 00:00:00+00:00 |  229500.0 |   3.0    |    1.0    |
| 3793500160 | 2015-03-12 00:00:00+00:00 |  323000.0 |   3.0    |    2.5    |
+------------+---------------------------+-----------+----------+-----------+
+-------------+----------+--------+------------+------+-----------+-------+------------+
| sqft_living | sqft_lot | floors | waterfront | view | condition | grade | sqft_above |
+-------------+----------+--------+------------+------+-----------+-------+------------+
|    1180.0   |   5650   |   1    |     0      |  0   |     3     |   7   |    1180    |
|    2570.0   |   7242   |   2    |     0      |  0   |     3     |   7   |    2170    |
|    770.0    |  10000   |   1    |     0      |  0   |     3     |   6   |    770     |
|    1960.0   |   5000   |   1    |     0      |  0   |     5     |   7   |    1050    |
|    1680.0   |   8080   |   1    |     0      |  0   |     3     |   8   |    1680    |
|    5420.0   |  101930  |   1    |     0      |  0   |     3     |   11  |    3890    |
|    1715.0   |   6819   |   2    |     0      |  0   |     3     |   7   |    1715    |
|    1060.0   |   9711   |   1    |     0      |  0   |     3     |   7   |    1060    |
|    1780.0   |   7470   |   1    |     0      |  0   |     3     |   7   |    1050    |
|    1890.0   |   6560   |   2    |     0      |  0   |     3     |   7   |    1890    |
+-------------+----------+--------+------------+------+-----------+-------+------------+
+---------------+----------+--------------+---------+-------------+
| sqft_basement | yr_built | yr_renovated | zipcode |     lat     |
+---------------+----------+--------------+---------+-------------+
|       0       |   1955   |      0       |  98178  | 47.51123398 |
|      400      |   1951   |     1991     |  98125  | 47.72102274 |
|       0       |   1933   |      0       |  98028  | 47.73792661 |
|      910      |   1965   |      0       |  98136  |   47.52082  |
|       0       |   1987   |      0       |  98074  | 47.61681228 |
|      1530     |   2001   |      0       |  98053  | 47.65611835 |
|       0       |   1995   |      0       |  98003  | 47.30972002 |
|       0       |   1963   |      0       |  98198  | 47.40949984 |
|      730      |   1960   |      0       |  98146  | 47.51229381 |
|       0       |   2003   |      0       |  98038  | 47.36840673 |
+---------------+----------+--------------+---------+-------------+
+---------------+---------------+-----+
|      long     | sqft_living15 | ... |
+---------------+---------------+-----+
| -122.25677536 |     1340.0    | ... |
|  -122.3188624 |     1690.0    | ... |
| -122.23319601 |     2720.0    | ... |
| -122.39318505 |     1360.0    | ... |
| -122.04490059 |     1800.0    | ... |
| -122.00528655 |     4760.0    | ... |
| -122.3270

#### Setting up SArray slices from SFrame, relevant aggregate values
* SArray: sqft_living
* SArray: sqft_living ^ 2
* SArray: price
* SArray: sqft_living * price
* Integer: number of rows in train_data
* Integer: sum of sqft_living
* Integer: sum of sqft_living ^ 2
* Integer: sum of price
* Integer: sum of sqft_living * price

In [5]:
# Setting up objects for each of the elements out of train_data
slice_sqft_living = train_data['sqft_living']
slice_sqft_living_squared = train_data['sqft_living'] * train_data['sqft_living']
slice_price = train_data['price']
slice_sqft_living_price_product = train_data['sqft_living'] * train_data['price']
num_rows = train_data.num_rows()

# Setting up objects for summarization of each new object from train_data
sum_sqft_living = slice_sqft_living.sum()
sum_sqft_living_squared = slice_sqft_living_squared.sum()
sum_price = slice_price.sum()
sum_sqft_living_price_product = slice_sqft_living_price_product.sum()

In [6]:
# Validating input slope
slope = (sum_sqft_living_price_product - ((sum_price * sum_sqft_living) / num_rows)) / (sum_sqft_living_squared - ((sum_sqft_living * sum_sqft_living) / num_rows))

In [7]:
# Validating intercept
intercept = (sum_price / num_rows) - (slope * (sum_sqft_living / num_rows))

In [8]:
print "Slope: " + str(slope)
print "Intercept: " + str(intercept)

Slope: 281.958838568
Intercept: -47116.0765749


**NOTE:** Specifically for train_data, I validated the above outputs using Excel against a CSV format of train_data
* Slope = 281.96
* Intercept = -47116

#### Setting up function that handles the above logic based on input feature and output, returns intercept and slope

In [9]:
def simple_linear_regression(input_feature, output):
    # Calculating products
    squared_input_feature = input_feature * input_feature
    product_input_feature_output = input_feature * output
    
    # Calculating aggregate values
    num_rows = input_feature.size()
    sum_input_feature = input_feature.sum()
    sum_squared_input_feature = squared_input_feature.sum()
    sum_output = output.sum()
    sum_product_input_output = product_input_feature_output.sum()
    
    # Calculating slope
    slope = (sum_product_input_output - ((sum_output * sum_input_feature) / num_rows)) / (sum_squared_input_feature - ((sum_input_feature * sum_input_feature) / num_rows))
    
    # Calculating intercept
    intercept = (sum_output / num_rows) - (slope * (sum_input_feature / num_rows))
    
    # Return intercept and slope
    return (intercept, slope)

#### Validating function produces same result as manual calculation for train_data['sqft_living'] and train_data['price']

In [10]:
sqft_intercept, sqft_slope = simple_linear_regression(train_data['sqft_living'], train_data['price'])
print "Slope: " + str(sqft_slope)
print "Intercept: " + str(sqft_intercept)

Slope: 281.958838568
Intercept: -47116.0765749


#### CONFIRMED! Function produces same values as manual mathematical approach

In [11]:
def get_regression_predictions(input_feature, intercept, slope):
    predicted_output = intercept + (slope * input_feature)
    return (predicted_output)

In [13]:
house_2650_sqft = get_regression_predictions(2650, sqft_intercept, sqft_slope)
house_2650_sqft

700074.8456294581

#### Creating RSS function
Must calculate the difference of output and predicted value for each input, square this difference, and summate the total column

In [17]:
def get_residual_sum_of_squares(input_feature, output, intercept, slope):
    
    # Calculate regression prediction for input_feature
    prediction = get_regression_predictions(input_feature, intercept, slope)
    squared_difference = (prediction - output) ** 2
    
    # Calculate RSS as sum of feature squared
    RSS = squared_difference.sum()
    
    return(RSS)

In [18]:
# Calculating RSS for the train_data['sqft_living'], train_data['price']
rss_sqft_intercept, rss_sqft_slope = simple_linear_regression(train_data['sqft_living'], train_data['price'])

rss_sqft_living = get_residual_sum_of_squares(train_data['sqft_living'], train_data['price'], rss_sqft_intercept, rss_sqft_slope)
rss_sqft_living

1201918356321968.0

In [21]:
# Calculating TSS, R-squared to confirm that R-squared ≈ 0.49409
tss_sqft_living = ((train_data['price'] - train_data['price'].mean()) ** 2).sum()
1 - (rss_sqft_living / tss_sqft_living)

0.4940913990347858

#### Creating inverse regression prediction to accept price, return predicted sqft_living

In [22]:
def inverse_regression_predictions(output, intercept, slope):
    estimated_feature = (output - intercept) / slope
    return(estimated_feature)

**TEST: Will we return approx 2650 sq feet based on the following price input: 700074.8456294581?**

In [23]:
inverse_regression_predictions(700074.8456294581, sqft_intercept, sqft_slope)

2650.0

In [33]:
inverse_regression_predictions(800000, sqft_intercept, sqft_slope)

3004.396247615945

**CONFIRMED! We have a functioning input feature estimator**

## Estimating price from bedrooms on train_data
For this, we will use train_data['price'] as our output, and train_data['bedrooms'] as our input feature. We can re-use our existing functions to easily solve this simple linear regression.

In [26]:
bedroom_intercept, bedroom_slope = simple_linear_regression(train_data['bedrooms'], train_data['price'])
print "Intercept: " + str(bedroom_intercept)
print "Slope: " + str(bedroom_slope)

Intercept: 109473.180469
Slope: 127588.952175


In [28]:
rss_bedrooms = get_residual_sum_of_squares(train_data['bedrooms'],
                                          train_data['price'],
                                          bedroom_intercept,
                                          bedroom_slope)
print "RSS for Bedroom feature: " + str(rss_bedrooms)

RSS for Bedroom feature: 2.14324449423e+15


In [29]:
rss_bedrooms > rss_sqft_living

True

## Estimating price for bedrooms on test_data
For this, we will use test_data['price'] as our output, and test_data['bedrooms'] as our input feature. We can re-use our existing functions to easily solve this simple linear regression.

In [35]:
'''
test_bedroom_intercept, test_bedroom_slope = simple_linear_regression(test_data['bedrooms'],
                                                                     test_data['price'])
print "Intercept on test_data: " + str(test_bedroom_intercept)
print "Slope on test_data: " + str(test_bedroom_slope)
'''
test_rss_sqft_living = get_residual_sum_of_squares(test_data['sqft_living'],
                                                  test_data['price'],
                                                  sqft_intercept,
                                                  sqft_slope)

print "RSS for Sqft_Living feature on test_data with train_data intercept, slope: " + str(test_rss_sqft_living)


test_rss_bedrooms = get_residual_sum_of_squares(test_data['bedrooms'],
                                               test_data['price'],
                                               bedroom_intercept,
                                               bedroom_slope)

print "RSS for Bedroom feature on test_data with train_data intercept, slope: " + str(test_rss_bedrooms)

print "Is RSS_bedrooms greater than RSS_sqft_living? " + str(test_rss_bedrooms > test_rss_sqft_living)

RSS for Sqft_Living feature on test_data with train_data intercept, slope: 2.75402936247e+14
RSS for Bedroom feature on test_data with train_data intercept, slope: 4.93364582868e+14
Is RSS_bedrooms greater than RSS_sqft_living? True
